clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -488,7 +491,19 @@ private void close(KafkaChannel channel) {
         } catch (IOException e) {
             log.error("Exception closing connection to node {}:", channel.id(), e);
         }
-        this.stagedReceives.remove(channel);
+
+        // Keep track of closed channels with pending receives so that all received records
+        // may be processed. For example, when producer with acks=0 sends some records and
+        // closes its connections, a single poll() in the broker may receive records and
+        // handle close(). Closed channels are retained until the current poll() processing
+        // completes to enable all records received on the channel to be processed.
+        Deque<NetworkReceive> deque = this.stagedReceives.remove(channel);
+        if (deque != null) {
+            while (!deque.isEmpty()) {
+                addToCompletedReceives(channel, deque);
```

Comment:
Hmm, not sure about this code. Currently, each selector.poll() is guaranteed to only return at most 1 receive for a given channel. This is important to guarantee that requests from the same channel are processed on the broker in the order they are sent. With this change, SocketServer can potentially put more than one outstanding request from a channel in the request queue. Since the requests could be handled by different request handler threads, there is no guarantee in which order those requests will be processed, which will broker our message ordering semantic.